vortex-array/src/operator/filter.rs

```diff
@@ -0,0 +1,155 @@
+use crate::operator::{
+    BatchBindCtx, BatchExecution, BatchExecutionRef, BatchOperator, Operator, OperatorId,
+    OperatorRef,
+};
+use crate::{Array, Canonical, IntoArray};
+
+#[derive(Debug)]
+pub struct FilterOperator {
+    child: OperatorRef,
+    mask: Mask,
+}
+
+impl PartialEq for FilterOperator {
+    fn eq(&self, other: &Self) -> bool {
+        self.child.eq(&other.child) && self.mask.eq(&other.mask)
+    }
+}
+impl Eq for FilterOperator {}
+
+impl Hash for FilterOperator {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.child.hash(state);
+        // Hash the discriminant first
+        std::mem::discriminant(&self.mask).hash(state);
+        match &self.mask {
+            Mask::AllTrue(len) => len.hash(state),
+            Mask::AllFalse(len) => len.hash(state),
+            Mask::Values(values) => {
+                Arc::as_ptr(values).hash(state);
+            }
+        }
+    }
+}
+
+impl FilterOperator {
+    pub fn new(child: OperatorRef, mask: Mask) -> FilterOperator {
+        assert_eq!(
+            child.len(),
+            mask.len(),
+            "Mask length must match child length"
+        );
+        FilterOperator { child, mask }
+    }
+
+    pub fn mask(&self) -> &Mask {
+        &self.mask
+    }
+}
+
+impl Operator for FilterOperator {
+    fn id(&self) -> OperatorId {
+        OperatorId::from("vortex.filter")
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn dtype(&self) -> &DType {
+        self.child.dtype()
+    }
+
+    fn len(&self) -> usize {
+        self.mask.true_count()
+    }
+
+    fn children(&self) -> &[OperatorRef] {
+        slice::from_ref(&self.child)
+    }
+
+    fn with_children(self: Arc<Self>, children: Vec<OperatorRef>) -> VortexResult<OperatorRef> {
+        Ok(Arc::new(FilterOperator {
+            child: children.into_iter().next().vortex_expect("missing child"),
+            mask: self.mask.clone(),
+        }))
+    }
+
+    fn reduce_children(&self) -> VortexResult<Option<OperatorRef>> {
+        // If none of the children are position-preserving, we cannot push down the filter.
+        if !self
+            .child
+            .children()
+            .iter()
+            .enumerate()
+            .any(|(i, child)| child.is_position_preserving(i).unwrap_or_default())
+        {
+            return Ok(None);
+        }
+
+        // We push down the filter operator to any child that is aligned to the parent.
+        let children = (0..self.child.nchildren())
+            .map(|i| {
+                let child = self.child.children()[i].clone();
+
+                if child.is_position_preserving(i).unwrap_or_default() {
+                    // Push-down the filter to this child.
+                    Arc::new(FilterOperator::new(child, self.mask.clone()))
+                } else {
+                    child
+                }
+            })
+            .collect();
+
+        Ok(Some(self.child.clone().with_children(children)?))
```

Comment:
this is not true `is_position_preserving == None` drops the filter on that child even if its unknown if its `is_position_preserving` need to bail out if this is undefined.